packages/playwright-core/src/server/socksClientCertificatesInterceptor.ts

```diff
@@ -84,51 +123,65 @@ class SocksProxyConnection {
         callback();
       }
     });
-    const dummyServer = tls.createServer({
-      key: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/key.pem')),
-      cert: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/cert.pem')),
-    });
-    dummyServer.emit('connection', this.internal);
-    dummyServer.on('secureConnection', internalTLS => {
-      internalTLS.on('close', () => this.socksProxy._socksProxy.sendSocketEnd({ uid: this.uid }));
-
-      const tlsOptions: tls.ConnectionOptions = {
-        socket: this.target,
-        host: this.host,
-        port: this.port,
-        rejectUnauthorized: !this.socksProxy.ignoreHTTPSErrors,
-        ...clientCertificatesToTLSOptions(this.socksProxy.clientCertificates, `https://${this.host}:${this.port}/`),
-      };
-      if (!net.isIP(this.host))
-        tlsOptions.servername = this.host;
-      if (process.env.PWTEST_UNSUPPORTED_CUSTOM_CA && isUnderTest())
-        tlsOptions.ca = [fs.readFileSync(process.env.PWTEST_UNSUPPORTED_CUSTOM_CA)];
-      const targetTLS = tls.connect(tlsOptions);
-
-      internalTLS.pipe(targetTLS);
-      targetTLS.pipe(internalTLS);
-
-      // Handle close and errors
-      const closeBothSockets = () => {
-        internalTLS.end();
-        targetTLS.end();
-      };
-
-      internalTLS.on('end', () => closeBothSockets());
-      targetTLS.on('end', () => closeBothSockets());
-
-      internalTLS.on('error', () => closeBothSockets());
-      targetTLS.on('error', error => {
-        const responseBody = 'Playwright client-certificate error: ' + error.message;
-        internalTLS.end([
-          'HTTP/1.1 503 Internal Server Error',
-          'Content-Type: text/html; charset=utf-8',
-          'Content-Length: ' + Buffer.byteLength(responseBody),
-          '\r\n',
-          responseBody,
-        ].join('\r\n'));
-        closeBothSockets();
+    alpnCache.get(rewriteToLocalhostIfNeeded(this.host), this.port, serverRequestedAlpnProtocol => {
+      debugLogger.log('socks', `Proxy->Target ${this.host}:${this.port} chooses ALPN ${serverRequestedAlpnProtocol}`);
+      const dummyServer = tls.createServer({
+        key: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/key.pem')),
+        cert: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/cert.pem')),
+        ALPNProtocols: serverRequestedAlpnProtocol === 'h2' ? ['h2', 'http/1.1'] : ['http/1.1'],
```

Comment:
I wonder whether we want to support `http/1.0`?

Reply:
I think most of the web is 1.1 nowadays - I'd hold off for now.